Add podman static build

saschagrunert opened the pull request:
We’re now able to build a static podman binary based on a custom nix
derivation. This is integrated in cirrus as well, whereas a later target
would be to provide a self-contained static binary bundle which can be
installed on any Linux x64-bit system.

Fixes: https://github.com/containers/libpod/issues/1399

Test it out yourself:
```bash
> make build-static
…
> ldd bin/podman-static
        not a dynamic executable
```

saschagrunert commented:
Just have to make the CI happy :)

vrothberg commented:
> This is integrated in cirrus as well, whereas a later target
would be to provide a self-contained static binary bundle which can be
installed on any Linux x64-bit system.

I feel rather objective toward that idea. There's a whole load of issues with static binaries working around NSS (breaking LDAP etc.). We don't provide static builds for that reason. If we allow for static builds, we need to make clear that this is not supported. I think we need some way in `podman info` to see if it's a static binary; issues should only be filed for dynamic ones.

TomSweeneyRedHat commented on Makefile:
```diff
@@ -213,6 +213,30 @@ bin/podman.cross.%: .gopathok
+	@nix run -f channel:nixpkgs-unstable nix-prefetch-git -c nix-prefetch-git \
+		--no-deepClone https://github.com/nixos/nixpkgs > nix/nixpkgs.json
+
+NIX_IMAGE ?= quay.io/crio/nix-podman:1.0.0
```
could we have this live in quay.io/podman/nix-podman:1.0.0 instead?

saschagrunert replied:
Yes, but I do not have write permissions to that repository :+1:

saschagrunert replied:
Fixed as discussed offline.

saschagrunert commented:
> > This is integrated in cirrus as well, whereas a later target
> > would be to provide a self-contained static binary bundle which can be
> > installed on any Linux x64-bit system.
> 
> I feel rather objective toward that idea. There's a whole load of issues with static binaries working around NSS (breaking LDAP etc.). We don't provide static builds of Skopeo for that reason. If we allow for static builds, we need to make clear that this is not supported. I think we need some way in `podman info` to see if it's a static binary; issues should only be filed for dynamic ones.

Do you recap which kind of issue that was? I'm not aware of any issues for example in conjunction with the CRI-O static binary bundle.

rhatdan commented:
Discussion here
https://github.com/containers/skopeo/pull/670

saschagrunert commented:
> Discussion here
> [containers/skopeo#670](https://github.com/containers/skopeo/pull/670)

Thanks. To me it looks like an usual bug in golang which has been fixed long time ago, right? I would not say that this falls into the category of: load of issues with static binaries. WDYT?

vrothberg commented:
> > Discussion here
> > [containers/skopeo#670](https://github.com/containers/skopeo/pull/670)
> 
> Thanks. To me it looks like an usual bug in golang which has been fixed long time ago, right? I would not say that this falls into the category of: load of issues with static binaries. WDYT?

I don't see that bypassing NSS has been resolved. @mtrmac will know.

rhatdan commented:
This is an issue of support.  It is likely that a static version of podman would not work well on a system that used nsswitch to modify the way host resolution worked.  Or how UID/GID lookups worked.
It would just fall back to reading /etc/hosts and /etc/resolv.conf, /etc/passwd, /etc/shadow.

This is definitely not something we would want to 'SUPPORT',   But I have no problem providing it for people to play with.

I agree that `podman info` should display that it was built staticly to prevent confusion.

saschagrunert commented:
> I agree that `podman info` should display that it was built staticly to prevent confusion.

Sounds good, I'll follow up with that on Monday if there are no further concerns.

mtrmac commented:
AFAIK the NSS situation remains the same (either dynamically link, or use a Go-only approximation of the default behavior, loosing all NSS plugins).

Separately, statically-building device-mapper and/or LVM, if nothing else, seems pretty scary. Is it safe to mix&match versions of libraries dealing with filesystems and user data on one system?

---

My first, fairly strong, instinct is to say ”no, get tested packages from your distribution” (and that includes Nix). Admittedly I’ve been living in the distribution world for ages, so I’m not at all neutral.

And we do see that whether one makes container builds or static builds difficult, one or the other tends to happen anyway just because users feel they need one of those. So it might be, after all, worthwhile to produce one of the two just to make sure most people use a known build instead of home-grown differently-patched one. Perhaps. I have no idea how prevalent are the _uses_ of the static/container builds, just that at least one person probably makes one.

saschagrunert commented:
I added an additional PodmanLinkmode field to the `podman info` output.

TomSweeneyRedHat commented:
Code LGTM, I'll leave it to others for the should/should not debate.  Personally I think it's fine as long as it's documented.

saschagrunert commented:
> I still have very mixed feelings about it and I frankly don't have time (or desire) to get into Nix, but as long as I don't have to maintain it, I am cool with it 
> 
> What I would love to have is some visible statement that the nix stuff is not supported. Which further makes me wonder why to offer something that is not supported but I don't want to be the party pooper and really want some more opinions.
> 
> @mheon @baude PTAL

:man_shrugging: I think if the majority does not want to have it then feel free to close it.

rhatdan commented:
Since this is an upstream project, support is best effort.  We can document the short comings of this version and state what does and does not work.  But if community wants it and is willing to deal with issues, then I have no problem doing it.

jwhonce commented:
@saschagrunert 

I have a request https://github.com/containers/libpod/issues/5765 points out `info` is already a bit slow.  Could you change `podmanLinkmode()` to `PodmanLinkMode()` and use a build tag to just report which mode?  You would need two files linkmode_dynamic.go and linkmode_static.go which are tagged to be mutually exclusive.  That way `info` can call and pay a minimal price.

rhatdan commented:
I just added podman-remote-static, which was needed for CRC,  @saschagrunert is this still something you are working on?

saschagrunert commented:
> I just added podman-remote-static, which was needed for CRC, @saschagrunert is this still something you are working on?

Yes I updated the PR and also the suggestion from @jwhonce. Should I state somewhere explicitly that the static binary is not supported? I mean if we not ship it and podman info returns the "static" linkmode we should be probably fine already, right?

vrothberg commented:
> Should I state somewhere explicitly that the static binary is not supported? 

Yes, please add a note in the README. If we don't make it explicit, users will not know.

saschagrunert commented on README.md:
```diff
@@ -180,3 +180,9 @@ Podman offers a Varlink-based API for remote management of containers.
 Varlink support is in maintenance mode, and will be removed in a future release.
 For more details, you can see [this blog](https://podman.io/blogs/2020/01/17/podman-new-api.html).
+
+## Static Binary Builds
+The Cirrus CI integration within this repository contains a `static_build` job
+which produces a static Podman binary for testing purposes. Please note that
+this binary is not officially supported with respect to feature-completeness
+and functionality and should be only used for testing.
```
Added the requested note here. :point_up:

vrothberg replied:
Great, thanks a lot!

rhatdan commented:
/lgtm
